### PR TITLE
Rename preferLowPowerToHighPerformance to powerPreference.

### PR DESCRIPTION
**Pull request by kenrussell**
Document that WEBGL_lose_context extension should destroy graphics
resources.

### CODE REVIEW

**Comment by kenrussell** [comment]
Partially addresses #2282 .

@grorg @jdashg @RafaelCintron @zhenyao @kainino0x in particular please review.

Thanks.

File: specs/latest/1.0/index.html [hunk 1]
```diff
@@ -723,14 +723,19 @@ <h2 class="no-toc">Table of contents</h2>
         The <code>WebGLContextAttributes</code> dictionary contains drawing surface attributes and
         is passed as the second parameter to <code>getContext</code>.
     </p>
-    <pre class="idl">dictionary <dfn id="WebGLContextAttributes">WebGLContextAttributes</dfn> {
+    <!-- The capitalization rules here match those of the crossOrigin
+    attribute, which was added to HTMLMediaElement on behalf of
+    WebGL. -->
+    <pre class="idl">enum PowerPreference { "default", "low-power", "high-performance" };
```
Comment (jdashg) [inline specs/latest/1.0/index.html]: I do sort of like "balanced" instead of "default", but "default" is fine.

File: extensions/WEBGL_lose_context/extension.xml [hunk 2]
```diff
@@ -37,13 +37,19 @@ interface WEBGL_lose_context {
...
+      <p>When this function is called and the context is not lost, simulate
       losing the context so as to trigger the steps described in the WebGL
       spec for handling context lost. The context will remain in the lost
       state according to the WebGL specification until
       <code>restoreContext()</code> is called.  If the context is already
       lost when this function is called, generate an
-      <code>INVALID_OPERATION</code> error.
+      <code>INVALID_OPERATION</code> error.</p>
+
+      <p>Implementations should destroy the underlying graphics context and
```
Comment (jdashg) [inline extensions/WEBGL_lose_context/extension.xml]: Thanks for adding this text, though it'd be nice to have it in a separate PR.

File: specs/latest/1.0/index.html (hunk 1 above)
Comment (kainino0x) [inline specs/latest/1.0/index.html]: I'm not sure if "balanced" describes what we'll be doing. On an ACPI system, "balanced" usually refers to automatic frequency scaling. If this mostly controls which GPU is used, "default" makes more sense.

**Comment by kainino0x** [comment]
Hm, any chance we can agree on meanings for all 3 values ("default", "low-power", "high-performance")? Specifically, the meaning of "default" is unclear to me.

File: specs/latest/1.0/index.html (hunk 1 above)
Comment (grorg) [inline specs/latest/1.0/index.html]: This needs to have a WebGL prefix.

`enum WebGLPowerPreference { ...`

File: specs/latest/1.0/index.html [hunk 3]
```diff
@@ -723,14 +723,19 @@ <h2 class="no-toc">Table of contents</h2>
...
+
+dictionary <dfn id="WebGLContextAttributes">WebGLContextAttributes</dfn> {
     GLboolean alpha = true;
     GLboolean depth = true;
     GLboolean stencil = false;
     GLboolean antialias = true;
     GLboolean premultipliedAlpha = true;
     GLboolean preserveDrawingBuffer = false;
-    GLboolean preferLowPowerToHighPerformance = false;
+    PowerPreference powerPreference = "default";
```
Comment (grorg) [inline specs/latest/1.0/index.html]: Should be WebGLPowerPreference.

**Comment by grorg** [comment]
The commit wording suggests that this change is just about the context lost extension, but the title is about the renaming of the attribute. The commit does a bit of both.

**Comment by grorg** [comment]
I'm not sure if this should be called `performancePreference` or `powerPreference`. I don't have strong opinions.

In my PR #2295 , I used the following wording.

performancePreference

 Provides a hint to the user agent indicating what configuration of GPU is suitable
for this WebGL context. This may influence which GPU is used in a system with
multiple GPUs. For example, a dual-GPU system might have one GPU that consumes less
power at the expense of rendering performance. Note that this property is only a hint and
a WebGL implementation may choose to ignore it. The allowed values are:

*default*

Let the user agent decide which GPU configuration is most suitable. This is
the default value.

*high-performance*

Indicates a request for a GPU configuration that values high
rendering performance over power consumption.
Any content that chooses this value should
be aware that it is much more likely that the user agent will
force a lost context at any time. Developers should always provide
event handlers for the context lost and
context restored events if they specify this value. Implementations may
decide to initially respect this request and, after some time, choose to
lose the context and restore a new context ignoring the request. Developers
are encouraged to only specify this value if they believe it is absolutely
necessary, since it may degrade the overall system performance and consume more
of the user's battery.

*low-power*

Indicates a request for a GPU configutation that values power consumption
over rendering performance. Generally this means the content is unlikely to be
constrained by drawing. For example, it may render only one frame a
second, it does not issuing draw calls with a lot of geometry or complex
shaders, or it is a very small HTML canvas element. Developers are encouraged
to specify this value if their content allows, since it may consume less
of the user's battery.

**Comment by kainino0x** [comment]
I'll move my comments from #2295 to this thread.

> _high-performance_

Overall I really like this wording, but I'm concerned that developers may read this as "you can either get high performance with low reliability or vice versa" which sounds scary and isn't accurate.

Specifically, I think we should make it clear that if a user is actively using "high" content, it shouldn't spontaneously lose context to save power (dunno how that's implemented). Here is also a good place to make it clear that recovering from a lost context isn't that hard - since the JS state is preserved.

> _low-power_

How about this slightly tweaked wording?

Indicates a request for a GPU configuration that prioritizes power saving over rendering performance. Generally, content should use this if it is unlikely to be constrained by drawing performance; for example, if it renders only one frame a second, draws only relatively simple geometry with simple shaders, or uses a very small HTML canvas element. Developers (...)

File: specs/latest/1.0/index.html (hunk 1 above)
Comment (grorg) [inline specs/latest/1.0/index.html]: Also, isn't this a pre class idl inside an existing pre class idl? No need to wrap it.

I suggest that the enum go one section above though, in the types section.

**Comment by RafaelCintron** [comment]
I am fine with wording that provides examples of instances where the browser may decide to lose-context on your canvas.  However, I think we should set clear expectations that context lost can happen no matter what, even when you ask for "low" and get "low".  On memory constrained machines like low end phones (which only have one GPU) dumping canvas contexts that use lots of memory is a good way to make room for other foreground applications or browser tabs.

**Comment by grorg** [comment]
I've landed the rename in WebKit, so I'm no longer proposing a name change :)
https://trac.webkit.org/changeset/212316

Now I'm starting on implementing the modes a bit better. @kenrussell, any thoughts on the proposed wording?

**Comment by kenrussell** [comment]
@grorg Working on incorporating the above feedback. Will push a new revision for review shortly.

File: extensions/WEBGL_lose_context/extension.xml (hunk 2 above)
Comment (kenrussell) [inline extensions/WEBGL_lose_context/extension.xml]: Undid this change; will submit it in a separate pull request.

**Comment by kenrussell** [comment]
@grorg @jdashg @RafaelCintron revised based on all of your feedback. Dean, in particular, thanks for your starting text; I edited it somewhat but used large portions.

Please re-review. Thanks.

File: specs/latest/1.0/index.html [hunk 4]
```diff
@@ -790,14 +797,60 @@ <h2 class="no-toc">Table of contents</h2>
...
+                  </p>
+                  <p>
+                    Let the user agent decide which GPU configuration is most suitable. This is the
+                    default value.
+                  </p>
+                  <p>
+                    <em>high-performance</em>
+                  </p>
+                  <p>
+                    Indicates a request for a GPU configuration that values high rendering
```
Comment (kainino0x) [inline specs/latest/1.0/index.html]: I would change "values" to "prioritizes" here as well.

Comment (kenrussell) [inline specs/latest/1.0/index.html]: Thanks for the suggestion. Done.

File: specs/latest/1.0/index.html [hunk 5]
```diff
@@ -790,14 +797,60 @@ <h2 class="no-toc">Table of contents</h2>
...
+                less power at the expense of rendering performance. Note that this property is
+                only a hint and a WebGL implementation may choose to ignore it.
+              </p>
+              <p>
+                WebGL implementations use context lost and restored events to regulate power and
+                memory consumption, <em>even if</em> this attribute is not specified. <em>The
+                application must have registered
+                <code>"webglcontextlost"</code> and <code>"webglcontextrestored"</code> event
+                listeners on the canvas element before fetching a WebGLRenderingContext specifying
+                this attribute, or "default" will be used.</em>
```
Comment (grorg) [inline specs/latest/1.0/index.html]: I like this rule. I thought we only talked about it informally, but I'm glad to see it in the specification.

File: specs/latest/1.0/index.html [hunk 6]
```diff
@@ -790,14 +797,60 @@ <h2 class="no-toc">Table of contents</h2>
...
+                  </p>
+                  <p>
+                    Let the user agent decide which GPU configuration is most suitable. This is the
+                    default value.
+                  </p>
+                  <p>
+                    <em>high-performance</em>
+                  </p>
+                  <p>
+                    Indicates a request for a GPU configuration that prioritizes high rendering
```
Comment (grorg) [inline specs/latest/1.0/index.html]: Maybe just "rendering performance" - removing "high"?

File: specs/latest/1.0/index.html [hunk 7]
```diff
@@ -790,14 +797,60 @@ <h2 class="no-toc">Table of contents</h2>
...
+                  </p>
+                  <p>
+                    <em>low-power</em>
+                  </p>
+                  <p>
+                    Indicates a request for a GPU configuration that prioritizes power saving over
+                    rendering performance. Generally, content should use this if it is unlikely to
+                    be constrained by drawing performance; for example, if it renders only one frame
+                    per second, draws only relatively simple geometry with simple shaders, or uses a
+                    small HTML canvas element. Developers are encouraged to specify this value if
```
Comment (grorg) [inline specs/latest/1.0/index.html]: "are encouraged to use this value", rather than "specify"?

File: specs/latest/1.0/index.html (hunk 5 above)
Comment (grorg) [inline specs/latest/1.0/index.html]: Actually, now that I think about this, the developer is free to add an event listener on any parent of the canvas element, up to the Document (and maybe even the window object).

So maybe the wording should be that there is a registered listener that would handle the lost/restore events if they were dispatched to the canvas element.

**Comment by kenrussell** [comment]
Thanks @grorg for your review and further suggestions. Incorporated them. Please re-review.

**Comment by kenrussell** [comment]
Thanks @grorg . Any other comments @jdashg @kainino0x @RafaelCintron ?

File: specs/latest/1.0/index.html [hunk 8]
```diff
@@ -790,14 +797,60 @@ <h2 class="no-toc">Table of contents</h2>
...
+                less power at the expense of rendering performance. Note that this property is
+                only a hint and a WebGL implementation may choose to ignore it.
+              </p>
+              <p>
+                WebGL implementations use context lost and restored events to regulate power and
+                memory consumption, <em>even if</em> this attribute is not specified. <em>The
+                application must have registered an event listener which would handle the
+                <code>"webglcontextlost"</code> and <code>"webglcontextrestored"</code> events if
+                they were dispatched to the canvas element before fetching a WebGLRenderingContext
+                specifying this attribute, or "default" will be used.</em>
```
Comment (kainino0x) [inline specs/latest/1.0/index.html]: I find this hard to parse, how about:

WebGL implementations use context loss/restoration events to regulate power and memory consumption, <em>regardless</em> of the value of this attribute. At WebGLRenderingContext creation time, this attribute will be ignored if no event listeners exist which would handle both <code>"webglcontextlost"</code> and <code>"webglcontextrestored"</code> events dispatched to the canvas element.

Comment (kenrussell) [inline specs/latest/1.0/index.html]: Thanks for the suggestion. Incorporated your text.

**Comment by kenrussell** [comment]
Further thoughts?

**Comment by RafaelCintron** [comment]
LGTM.

**Comment by kenrussell** [comment]
Thanks everyone. Merging now.

**Comment by grorg** [comment]
Awesome. Thanks @kenrussell . WebKit's implementation will land over the next few days.
